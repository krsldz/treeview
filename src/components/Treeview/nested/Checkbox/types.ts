import { ChangeEvent } from "react";
import { CustomClassNames } from "../../types";

export type CheckboxProps = {
  label: string;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  checked: boolean;
  disabled?: boolean;
  indeterminate?: boolean;
  classNames: Partial<CustomClassNames>;
};
