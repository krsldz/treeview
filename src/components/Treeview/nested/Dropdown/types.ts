export type DropdownProps = { open: boolean; onClick: () => void };
