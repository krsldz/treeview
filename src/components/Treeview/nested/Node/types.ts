import { IconComponents } from "./../../types";
import { ReactNode } from "react";
import { CustomClassNames, NodeItem, Value } from "../../types";

export type NodeProps = {
  data: NodeItem;
  onlyRead: boolean;
  withIcons: boolean;
  selected: boolean;
  renderChildren: (children: Value[]) => ReactNode[];
  onToggle: (item: Value) => void;
  onChange: (node: NodeItem) => void;
  indeterminate?: boolean;
  classNames: Partial<CustomClassNames>;
  iconsComponents: IconComponents;
};
