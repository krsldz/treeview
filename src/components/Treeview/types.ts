export type Value = string | number;

export type EntityType = {
  id: Value;
  name: string;
};

export type NodeData = {
  title: string;
  value: Value;
  children?: NodeData[];
  id?: Value;
};

export type NodeItem = {
  data: NodeData;
  children: Value[];
  id: Value;
  parent: Value | null;
  open: boolean;
};

export type NodeMap = Record<string, NodeItem>;

export type TreeViewProps = {
  data: Array<NodeData>;
  onlyRead?: boolean;
  value?: EntityType[];
  onChange?: (value: EntityType[]) => void;
};
