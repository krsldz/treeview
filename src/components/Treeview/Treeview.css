html, body {
  font-family: sans-serif;
  font-size: 16px;
}

.wrap {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.storyCheckbox input[type="checkbox"] {
  appearance: none;
  -webkit-appearance: none;
  width: 20px;
  height: 20px;
  border: 2px solid #4caf50;
  border-radius: 50%;
  outline: none;
  cursor: pointer;
  position: relative;
  transition: background-color 0.2s ease, border-color 0.2s ease;
}

.storyCheckbox input[type="checkbox"]:checked {
  background-color: #4caf50;
  border-color: #4caf50;
  position: relative;
}

.storyCheckbox input[type="checkbox"]:checked::before {
  content: "";
  position: absolute;
  left: 5px;
  top: 0;
  width: 5px;
  height: 10px;
  border: solid white;
  border-width: 0 2px 2px 0;
  transform: rotate(45deg);
}

input.indeterminate {
  background-color: #4caf50;
  position: relative;
}

input.indeterminate::before {
  content: url("/public/indeterminate.svg");
  color: #f8f9fa;
  position: absolute;
  right: 3px;
  top: -4px;
}


.storyLabel {
  margin-left: 5px;
  font-weight: 200;
}

.storyDropdown {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: 2px solid #4caf50;
  background-color: #4caf50;
  color: #4caf50;
  cursor: pointer;
  transition: transform 0.2s ease, background-color 0.2s ease;
  font-size: 14px;
  line-height: 1;
  padding: 0;
}

.storyDropdown img {
  stroke: #ffff;
}
