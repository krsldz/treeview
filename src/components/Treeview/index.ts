export { default } from "./Treeview";

export { Value, EntityType, NodeData, IconComponents } from "./types";
