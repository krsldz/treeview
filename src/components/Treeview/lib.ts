import { EntityType, NodeData, NodeItem, NodeMap, Value } from "./types";

/**
 * Form a nodes object
 * @param {NodeData[]} nodes - data array
 * @param {Value | null}  parentId - parent element id
 * @param {Value[]} values - chosen values array
 * @returns {NodeMap}
 */
export const getNodeMap = (
  nodes: NodeData[],
  parentId: Value | null,
  values: Value[]
): NodeMap => {
  return nodes.reduce((map, node) => {
    const { value, children, id } = node;
    const open =
      children?.some((option) => values.includes(option.value)) ||
      values.includes(value);

    const treeNode: NodeItem = {
      children:
        node.children?.map((option) =>
          option.id ? option.id : option.value
        ) ?? [],
      data: node,
      id: id || value,
      open,
      parent: parentId || null,
    };

    let childrenMap;

    if (children) childrenMap = getNodeMap(children, treeNode.id, values);

    return { ...map, ...childrenMap, [treeNode.id]: treeNode };
  }, {} as NodeMap);
};

/**
 * Set up an array of current node id and its children elements
 * @param {NodeMap} map - a map of all nodes
 * @param {Value}  id - id of current node
 * @returns {Value[]}
 */
export const getNodeValues = (map: NodeMap, id: Value): Value[] => {
  const { children, data } = map[id];

  return [
    data.value,
    ...children.reduce(
      (values, id) => [...values, ...getNodeValues(map, id)],
      [] as Value[]
    ),
  ];
};

/**
 * Set up an array of unique values
 * @param {EntityType[]} arr - an array of chosen non-unique values
 * @returns {EntityType[]}
 */
export const getUniqueValues = (arr: EntityType[]): EntityType[] => {
  return [
    ...new Set(
      arr.map((item) => {
        const sortedObjectKeys = Object.keys(item).sort();
        const obj = Object.assign(
          {},
          ...sortedObjectKeys.map((i) => ({
            [i]: item[i as keyof typeof item],
          }))
        );
        return JSON.stringify(obj);
      })
    ),
  ].map((s) => JSON.parse(s));
};
