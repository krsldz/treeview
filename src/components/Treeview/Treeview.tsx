import React, {
  FC,
  memo,
  useCallback,
  useLayoutEffect,
  useMemo,
  useState,
} from "react";
import { EntityType, NodeItem, NodeMap, TreeViewProps, Value } from "./types";
import Node from "./nested/Node";
import { getNodeMap, getNodeValues, getUniqueValues } from "./lib";

import "./Treeview.css";

const Treeview: FC<TreeViewProps> = ({
  data,
  onChange,
  value = [],
  onlyRead = true,
}) => {
  const [nodeMap, setNodeMap] = useState<NodeMap>({});

  const selected = useMemo(() => {
    return value.map((i) => i.id);
  }, [value]);

  useLayoutEffect(() => {
    setNodeMap(getNodeMap(data, null, selected));
  }, []);

  const roots = useMemo(() => {
    return Object.keys(nodeMap).filter((key) => nodeMap[key].parent === null);
  }, [nodeMap]);

  const onToggle = useCallback(
    (id: Value): void => {
      const newNode = { ...nodeMap[id], open: !nodeMap[id].open };

      setNodeMap({ ...nodeMap, [id]: newNode });
    },
    [nodeMap]
  );

  const handleChange = useCallback(
    (node: NodeItem): void => {
      if (!onChange) return;

      const { data, children, parent, id } = node;

      let ids = [data.value];
      let returnValue: EntityType[] = [{ id: data.value, name: data.title }];

      if (children) {
        ids = getNodeValues(nodeMap, id);
      }

      const childrenItems = data?.children?.map((i) => {
        return { id: i.value, name: i.title };
      });

      returnValue = childrenItems
        ? [...childrenItems, { id: data.value, name: data.title }]
        : [{ id: data.value, name: data.title }];

      const newValue = selected.includes(data.value)
        ? value.filter((item) => !ids.includes(item.id) && item.id !== parent)
        : getUniqueValues([...value, ...returnValue]);

      if (parent) {
        const parentNode = nodeMap[parent];
        const newValueIds = newValue.map((val) => val.id);
        const allChildrenSelected = parentNode.data.children?.every((child) =>
          newValueIds.includes(child.value)
        );
        if (allChildrenSelected) {
          returnValue = [
            { id: parentNode.data.value, name: parentNode.data.title },
          ];
          onChange([...newValue, ...returnValue]);
        } else {
          onChange(newValue);
        }
      } else {
        onChange(newValue);
      }
    },
    [value, nodeMap]
  );

  const renderNode = (id: Value) => {
    const node = nodeMap[id];
    const { data } = node;
    const isSelected = selected.includes(data.value);

    return (
      <Node
        key={id}
        data={node}
        onlyRead={onlyRead}
        renderChildren={(ids) => ids.map(renderNode)}
        onToggle={onToggle}
        onChange={handleChange}
        selected={isSelected}
      />
    );
  };

  return <div className="wrap">{roots.map(renderNode)}</div>;
};

export default memo(Treeview);
