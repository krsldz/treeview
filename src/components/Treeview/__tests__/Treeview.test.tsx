import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Treeview from "../Treeview";
import { TreeviewProps } from "../types";

describe("Treeview component", () => {
  let user: ReturnType<typeof userEvent.setup>;
  let treeviewProps: TreeviewProps;

  beforeEach(() => {
    user = userEvent.setup();

    treeviewProps = {
      onlyRead: false,
      data: [
        {
          children: [
            { title: "Option 1-1", value: 11 },
            { title: "Option 1-2", value: 12 },
          ],
          title: "Option 1",
          value: 1,
        },
        {
          children: [
            { title: "Option 2-1", value: 21 },
            { title: "Option 2-2", value: 22 },
          ],
          title: "Option 2",
          value: 2,
        },
      ],
      onChange: jest.fn(),
      value: [{ id: 1, name: "Option 1" }],
    };
  });

  it("renders correctly", () => {
    const { asFragment } = render(<Treeview {...treeviewProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("selects an item when it is clicked", async () => {
    const { getByRole } = render(<Treeview {...treeviewProps} />);

    await user.click(getByRole("checkbox", { name: "Option 1-2" }));

    expect(treeviewProps.onChange).toHaveBeenCalledWith([
      {
        id: 1,
        name: "Option 1",
      },
      {
        id: 12,
        name: "Option 1-2",
      },
    ]);
  });

  it("deselects an item when it is clicked again", async () => {
    const { getByRole } = render(<Treeview {...treeviewProps} />);

    const checkbox = getByRole("checkbox", { name: "Option 1" });

    await user.click(checkbox);

    await user.click(checkbox);

    expect(treeviewProps.onChange).toHaveBeenCalledWith([]);
  });

  it("selects all child items when the parent is selected", async () => {
    const { getByRole } = render(<Treeview {...treeviewProps} />);

    await user.click(getByRole("checkbox", { name: "Option 2" }));

    expect(treeviewProps.onChange).toHaveBeenCalledWith([
      { id: 1, name: "Option 1" },
      { id: 21, name: "Option 2-1" },
      { id: 22, name: "Option 2-2" },
      { id: 2, name: "Option 2" },
    ]);
  });

  it("selects the parent when all of its children are selected", async () => {
    const { getByRole } = render(
      <Treeview {...treeviewProps} value={[{ id: 11, name: "Option 1-1" }]} />
    );

    await user.click(getByRole("checkbox", { name: "Option 1-2" }));

    expect(treeviewProps.onChange).toHaveBeenCalledWith([
      { id: 11, name: "Option 1-1" },
      { id: 12, name: "Option 1-2" },
      { id: 1, name: "Option 1" },
    ]);
  });
});
