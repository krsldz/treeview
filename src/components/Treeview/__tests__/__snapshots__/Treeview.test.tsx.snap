// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Treeview component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="wrap"
  >
    <div
      class="nodeWrap"
    >
      <div
        class="dropdown"
      >
        <img
          alt="dropdown"
          class="icon"
          src="/chevronDown.svg"
        />
      </div>
      <div
        class="node"
      >
        <div
          class="checkboxWrap"
        >
          <input
            aria-label="Option 1"
            checked=""
            class=""
            type="checkbox"
          />
          <span>
            Option 1
          </span>
        </div>
        <div>
          <div
            class="nodeWrap"
          >
            <div
              class="node"
            >
              <div
                class="checkboxWrap title"
              >
                <input
                  aria-label="Option 1-1"
                  class=""
                  type="checkbox"
                />
                <span>
                  Option 1-1
                </span>
              </div>
            </div>
          </div>
          <div
            class="nodeWrap"
          >
            <div
              class="node"
            >
              <div
                class="checkboxWrap title"
              >
                <input
                  aria-label="Option 1-2"
                  class=""
                  type="checkbox"
                />
                <span>
                  Option 1-2
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="nodeWrap"
    >
      <div
        class="dropdown"
      >
        <img
          alt="dropdown"
          class="icon"
          src="/chevronLeft.svg"
        />
      </div>
      <div
        class="node"
      >
        <div
          class="checkboxWrap"
        >
          <input
            aria-label="Option 2"
            class=""
            type="checkbox"
          />
          <span>
            Option 2
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
